### ISSUE TITLE
user story no 9

### ISSUE THREAD

**Issue by rcapell** [switchonproject/sip-html5#46]
As a sip database maintainer confronted with today's heterogeneous online data resources, I have to be able to upload multiple spatial locations for a single data source, so I can provide a better representation for a data set compared to the bounding box area. 

Example GRDC: we cannot put each station as separate source into the sip, but station location points are an improvement compared to a global bounding box.

@duropa: please confirm that this is useful, or suggest changes.

**Comment by duropa** [comment]
this is useful,

but here i would like to note that, potentially, experiments in vwsl will bring some additional (new) results for those spatial locations.

will those be stored separately (spatial location will be duplicated, but the data will be different) or we will add results to existing spatial locations?this relates to the grdc dataset...

 
d.

 
 

 
-----Original message-----

**Comment by rcapell** [comment]
That's actually a good question, which also applies to other data sets as well. Any station or site can have multiple measurements which are potentially hosted in different data sets.
My guess after Friday's discussion is that it would be easiest to add a new spatial location for each data set.

**Comment by p-a-s-c-a-l** [comment]
> but here i would like to note that, potentially, experiments in vwsl will bring some additional (new) results for those spatial locations.

Since this is re-purposed data (a new experiment result) it should also become a new resource in the Meta-Data repository and the relationship to the source data should be described (e.g. Lineage: This data has been derived from ...)

> Any station or site can have multiple measurements which are potentially hosted in different data sets.

In this case, an **additional representation** can be added to the resource, so that a resource in the Meta-Data Repository can point to different data sets as shown in the following diagram:
![switch-on d2 6 diagrams](https://cloud.githubusercontent.com/assets/1788226/6213562/3a24c77e-b5f1-11e4-9bfb-45af3bcd36a1.png)

**Comment by p-a-s-c-a-l** [comment]
@rcapell Could you please provide some sample resources and the respective coordinates WKT format (e.h. MULTIPOINT or MULTIPOLYGON)?

**Comment by rcapell** [comment]
Pascal, I won't manage to do this this week but will prepare and send
something on Monday. I might come back for more info about the WKT format...

**Comment by rcapell** [comment]
I created a multipoint WKT (csv) file of GRDC stations, ie global runoff data collection. Will send by email to you now, @p-a-s-c-a-l 
Then you can comment if this works for you. There is no further information attached to each point, e.g. gauging site ID or name. This information was stripped when IU converted from point to multipoint.

**Comment by p-a-s-c-a-l** [comment]
@rcapell, thanks for the WKT!  This is how it looks on the map:

![screen shot 2015-03-02 at 17 02 39](https://cloud.githubusercontent.com/assets/1788226/6444532/6fed9148-c0fe-11e4-8623-b1630057d89e.png)

@mscholl will create an entry for the GRDC dataset in the meta-data repository and then you can test the spatial search. 

Then we have to decide if this approach is technically feasible, since showing so many points in map makes the client extremely slow.

**Comment by mscholl** [comment]
Hi, I created a GRDC that has the Multi Point geometry as sent. The resource can be found as expected:
![screen shot 2015-03-02 at 17 59 17](https://cloud.githubusercontent.com/assets/973421/6445702/1e269e06-c106-11e4-85d3-35313d8f9bd2.png)
You will have to use the "Intersect" setting and you will have to use the CUAHSI keyword "water" to actually find it as there are too many resources that have a world geometry which are found also if not limiting to "water". The resource can be visualised on the map, however, *I DO NOT RECOMMEND TO SWITCH TO THE MAP FOR RESULT VISUALISATION*. It will work but it will take very much time and space which might render the browser or even your OS unresponsive for quite some time, depending on the power of your computer. Mine is quite decent and it took roughly 2 min until I saw the result.

Additionally, besides the technical difficulties with such an approach, I don't think that it makes sense to split a resource that apparently is a "world" resource in thousands of points as the "search behaviour" will hardly change because the points effectively cover all of world, at least in this particular case. It would make more sense to split a "big resource" into smaller and more meaningful ones.

Bottom line: While technically possible this approach comes with some severe downsides and IMHO is hardly of any use for anybody.

**Comment by rcapell** [comment]
hm, I can't comment much on the technical feasibility, but I can say that it does make sense to have points instead of a world polygon from a hydrological perspective. A hydrological study could be performed in a catchments covering a few square km and if you search at that scale, GRDC covers only a tiny fraction of the earth.
Splitting into regions would principally not be a problem, but I wonder if one can connect multiple spatial sets (= WKT files) to one dataset?

**Comment by p-a-s-c-a-l** [comment]
Connecting multiple spatial sets (= WKT files) to one dataset (resource) is not possible both for technical and compatibility reasons. Therefore the GRDC database  should be spilt into meaningful smaller chunks (WMO Sub-Region, Country, ...) that can be registered as distinct resources in the SIP. Those resources can be assigned one category (GRDC)  like it has been done done for ASMR data.

Another issue is that GRDC forbids to provide a **direct download link** to GRDC data, neither to the whole database nor to the smaller chunk. It seems that you always have to fill in some kind of order form to get access to the real data: [Steps to Order River Discharge Time Series](http://www.bafg.de/GRDC/EN/02_srvcs/21_tmsrs/riverdischarge_node.html). So *all* GRDC resources in the SIP would point to the same url.

**Comment by rcapell** [comment]
Regarding multple spatial sets: yes, I thought I remembered from Saarbruecken that this was not possible. I prepared another WKT file with multipoints split by country and will email it to @p-a-s-c-a-l directly.
Regarding no direct downloads: that is true and i fear will be true for many other data sources as well. Nothing to do about that immediately, but maybe a successful sip will lead to an increase in requests that will eventually pressure providers into providing something more accessible?

**Comment by mscholl** [comment]
I implemented the possibility to have a WMS Layer representation for objects (#57). If you now perform the search like mentioned above (in the dev snapshot) you will see the points on the map. The default point sld is applied, though, that doesn't look that good but that's a minor issue....

**Comment by mscholl** [comment]
Well, I just talked to Pascal again and I should have read your concern more carefully: You want to search for a resource that is available WITHIN a specified search area, or to use the (mathematically) correct term you want the resource to be found to be ENCLOSED by the specified search area. If this is correct then the proposed approach will not work using any library for geoprocessing. To clarify:

You have this:
![multipoint-issue-1](https://cloud.githubusercontent.com/assets/973421/6467284/81bbb952-c1cd-11e4-8c78-a8fae01a297e.png)
And you want to search like so:
![multipoint-issue-2](https://cloud.githubusercontent.com/assets/973421/6467295/906862e8-c1cd-11e4-825b-81f7bdf82d1f.png)
Then the resource you are searching for is NOT ENCLOSED
![multipoint-issue-3](https://cloud.githubusercontent.com/assets/973421/6467382/1401cf90-c1ce-11e4-8ad2-534df95c8e85.png)

It would be ENCLOSED if you searched like this
![multipoint-issue-4](https://cloud.githubusercontent.com/assets/973421/6467372/fcbef614-c1cd-11e4-9962-28893c27716e.png)

which defeats the purpose. So what you want is mathematically not possible. On the other hand if you use INTERSECTION you will find the resource that you are asking for, however, you will also find every other resource that e.g. has the extend "WORLD".

**Comment by rcapell** [comment]
@mscholl thanks for the clarification. Lets see if I understood them :-) :
I am not quite sure if we at some point said that we should limit search to ENCLOSURE (but it is possible). My main point is that it is no use to get a result when there is no gauging station, because it is a point information. If a multipoint can only be found by finding all points belonging to it (the area enclosed by the dashed circle in your figures), then it is probably not the right choice. 
This is how I thought a search should work:
![discussion_search](https://cloud.githubusercontent.com/assets/6536193/6469542/ac63b384-c1dd-11e4-8382-91ab76da3275.png)
If I understand the multipoint structure correctly, then there is no way of having separate attributes, e.g. a station ID, for each point in a multipoint set. That is quite limiting for data resources like the GRDC where users need these IDs to fill in their order forms. On the other hand separate points for each station, all belonging to the same data resource, are not possible in your current database structure, correct? A replication of the data source for each station does not seem very useful either.
So, I am not quite sure how to move forward from here.

Regarding entries with "WORLD" extent: I can't think of a scenario where that could be useful at all in the context of a spatial search.

**Comment by p-a-s-c-a-l** [comment]
> I am not quite sure if we at some point said that we should limit search to ENCLOSURE (but it is possible).

The search is limited to enclosure, because all resources in the Meta-Data Repository have a bbox with world extent. So a spatial search with intersects is totally useless as it always returns all resources unless you specify an additional filter to limit the search results, e.g. text:"GRDC". That's a well know [issue](https://github.com/switchonproject/cids-custom-switchon/issues/61).

On the other hand, geospatial intersects doesn't seem to work well with Multipoints. The geospatial search is extremely slow (>30 seconds) and nearly unusable! For testing, I've chosen a bbox where no GRDC station is located plus an additional "GRDC" text filter to exclude any other resource. Sometimes resources are found, sometimes not. Probably there is discrepancy between the GRDC Stations layer that @mscholl has created (based on the stations KMZ from the GRDC website) and the Multipoint you've sent us. In any case, I don't think that the Multipoint approach should be further pursued.

> On the other hand separate points for each station, all belonging to the same data resource, are not possible in your current database structure, correct?

What would be the advantage if that would be possible? What is the expected result of a search then? You'll still get the complete resource. Or would you expected to get a list of matching stations? Imho this is then the same as "A replication of the data source for each station".

> Regarding entries with "WORLD" extent: I can't think of a scenario where that could be useful at all in the context of a spatial search.

Yes, but that's the current situation in the Meta-DB and I don't that this is going to change. Better meta-data is simply not available.

**Comment by rcapell** [comment]
> On the other hand separate points for each station, all belonging to the same data resource, are not   possible in your current database structure, correct?
> What would be the advantage if that would be possible? What is the expected result of a search then? You'll still get the complete resource. Or would you expected to get a list of matching stations? Imho this is then the same as "A replication of the data source for each station".

I thought that one could get a station ID as a result, instead of jsut the data source.

**Comment by p-a-s-c-a-l** [comment]
Ok, then we have to add each station as separate resource to the 
meta-db. Meta-Data for stations is available, we just have to import them.

Am 04.03.2015 um 10:15 schrieb rcapell:

**Comment by p-a-s-c-a-l** [comment]
This video shows the implementation of user story no 9:

[![user-story-9](https://cloud.githubusercontent.com/assets/1788226/6555941/24a27e12-c66a-11e4-80a0-7a47bfbb82ad.png)](http://youtu.be/vTQD86pj_fQ)

Please note that the GRDC Stations (red dots) shown on the map come from a separate Web Map Service. This is suboptimal, but the web mapping component cannot handle a MULTIPOINT with 9000+ points.

**Comment by duropa** [comment]
pascal wrote:
Ok, then we have to add each station as separate resource to the meta-db. Meta-Data for stations is available, we just have to import them.

I would be happy with that as we would like to add some additional data from our experiment to many of the grdc stations...(e.g. when the floods occur at the station, etc...)

**Comment by p-a-s-c-a-l** [comment]
@duropa @rcnijzink @rcnijzink 

You can test the new functionality in the [new development release](http://switchon.cismet.de/sip-snapshot/#/map) by searching for *GRDC* with *intersects* geo filter and a *polygon* as shown in the screen shot below: 

![grdc](https://cloud.githubusercontent.com/assets/1788226/6598792/54050df8-c806-11e4-8053-90e8836910e4.png)

Please note that the GRDC Stations shown on the map as red/yellow points come from dedicated Web Map Service at deltares. Sometimes the map layer is not refreshed properly. This is an issue we are already aware of.

**Comment by duropa** [comment]
it works well for me (even in firefox)...and it is really a nice step forward...
just one note...would it be possible to set the intersect as default?
and i still miss the tooltip text for the map and table results...

**Comment by p-a-s-c-a-l** [comment]
> it works well for me (even in firefox)...and it is really a nice step forward...

Thanks!

> just one note...would it be possible to set the intersect as default?

Yes I, can make this the default. However, whether this will improve spatial search or not depends on the result of the "[improve spatial extent of resources in the Meta-Data Repository](https://github.com/switchonproject/sip-html5/issues/63)" task. At the moment we have mainly world boxes in the Meta-Data Repository. Therefore a spatial intersects search will always return nearly all resources,  unless you make further restrictions (e.g. keywords).

> and i still miss the tooltip text for the map and table results

Yes, that's still on my [todo list](https://github.com/switchonproject/sip-html5/issues/58) ... ;)
